boot/BootFromDevice.c

```diff
@@ -0,0 +1,364 @@
+ *                                                                         *
+ *   This program is free software; you can redistribute it and/or modify  *
+ *   it under the terms of the GNU General Public License as published by  *
+ *   the Free Software Foundation; either version 2 of the License, or     *
+ *   (at your option) any later version.                                   *
+ *                                                                         *
+ *   Copyright (C) 2019 Stanislav Motylkov                                 *
+ *                                                                         *
+ ***************************************************************************/
+#include "boot.h"
+#include "video.h"
+#include "memory_layout.h"
+#include <shared.h>
+#include <filesys.h>
+#include "rc4.h"
+#include "sha1.h"
+#include "BootFATX.h"
+#include "xbox.h"
+#include "BootFlash.h"
+#include "cpu.h"
+#include "BootIde.h"
+#include "BootParser.h"
+#include "config.h"
+#include "iso_fs.h"
+
+/* GRUB request size should be either 256+4 or 265+4 based on previous code
+ * but I have no idea which is correct and why, but 256+4 makes more sense */
+#define GRUB_REQUEST_SIZE 256+4
+
+char* InitGrubRequest(int size, int drive, int partition) {
+	char* szGrub;
+
+	szGrub = (char*)malloc(size);
+	memset(szGrub, 0, size);
+
+	szGrub[0] = 0xff;
+	szGrub[1] = 0xff;
+	szGrub[2] = partition;
+	szGrub[3] = drive;
+
+	errnum = 0;
+	boot_drive = 0;
+	saved_drive = 0;
+	saved_partition = 0x0001ffff;
+	buf_drive = -1;
+	current_partition = 0x0001ffff;
+	current_drive = drive;
+	buf_drive = -1;
+	fsys_type = NUM_FSYS;
+	disk_read_hook = NULL;
+	disk_read_func = NULL;
+
+	return szGrub;
+}
+
+void FillConfigEntries(CONFIGENTRY* config, enum BootTypes bootType, int drive, int partition) {
+	CONFIGENTRY* currentConfigItem;
+
+	for (currentConfigItem = (CONFIGENTRY*) config; currentConfigItem != NULL; currentConfigItem = (CONFIGENTRY*)currentConfigItem->nextConfigEntry) {
+		currentConfigItem->bootType = bootType;
+
+		// Set the drive ID and partition IDs for the returned config items
+		switch (bootType) {
+			case BOOT_NATIVE:
+				currentConfigItem->drive = drive;
+				currentConfigItem->partition = partition;
+				break;
+			case BOOT_FATX:
+				break;
+			case BOOT_CDROM:
+				currentConfigItem->drive = drive;
+				break;
+		}
+	}
+}
+
+CONFIGENTRY* AddNestedConfigEntry(CONFIGENTRY* config, CONFIGENTRY* nested, char* title) {
+	CONFIGENTRY* root;
+
+	if (config != NULL)
+	{
+		/* Append new config entry to the end of list */
+		root = config;
+		while (config->nextConfigEntry != NULL) config = config->nextConfigEntry;
+		config->nextConfigEntry = (CONFIGENTRY*)malloc(sizeof(CONFIGENTRY));
+		config = config->nextConfigEntry;
+	}
+	else
+	{
+		/* No config entry, create new one */
+		root = (CONFIGENTRY*)malloc(sizeof(CONFIGENTRY));
+		config = root;
+	}
+	memset(config, 0x00, sizeof(CONFIGENTRY));
+	strncpy(config->title, title, strlen(title));
```

Comment:
The 3rd parameter should be the maximum number of bytes to write to destination (i.e. buffer size of `config->title`), not the length of the source string.